ntp.go

```diff
@@ -407,78 +409,57 @@ func parseTime(m *msg, recvTime ntpTime) *Response {
 
 	// Calculate values depending on other calculated values
 	r.RootDistance = rootDistance(r.RTT, r.RootDelay, r.RootDispersion)
-	r.CausalityViolation = causalityViolation(r.RTT, r.ClockOffset)
```

Comment:
I'm sad to see you dropping `CausalityViolation` when it's used in [prometheus/node_exporter](https://github.com/prometheus/node_exporter/blob/master/collector/ntp.go#L155) :-/

Reply:
I'm happy to add it back in.  I'd just like to have a better understanding of what it's trying to do.  How does one interpret the resulting value?  When is a value considered "bad"?

Reply:
I'll try do draw some diagram explaining this value.

Reply:
```
S: --[t1]-[t2]---------->
       \   \__ 
        \     \__ 
         \       \   
C: ------[t0]----[t3]--->
```
That's causality violation case. If clocks are in-sync then T1 is always greater then T2.
There is Offset that is estimate of clock error that assumes that network path is symmetric in terms of latency.
And CausalityViolation is lower bound of clock error, if T0 is greater than T1 then clocks are **at least** wrong by the difference and that's 100% true :)

```
S: ------[t1]-[t2]--->
         _/    |
       _/      | 
      /       /    
C: --[t0]----[t3]---->
```
That's another example. If we assume that server clock is "true" clock then packet arrives "earlier" than departs, so we have another esimate for minimum error value.

These error values are useful to check if local kernel wall clock is synced to local NTP daemon clock in case of `chrony` daemon or external NTP server providing "reference" clock. Offset is always non-zero due to network fluctuations and CausalityViolation may give lower-bound estimate in case of unsync clock.

I hope it helps.

Reply:
Ok, that adds clarity.

So what you're saying is that if `CV > 0`, then the client and server clocks are not just out of sync (i.e., `ClockOffset != 0`); they are also returning values that really should be impossible.  Is that correct?

Reply:
> are not just out of sync (i.e., `ClockOffset != 0`);

Estimated offset (ClockOffset) is always non-zero, but 5ms offset estimated using nearby server (1ms away) will show some CV, while remote server (RTT~100ms) will show CV=0 as there is some uncertainty due to processing time.

> I don't really understand how the "causality violation" happens in the first place

Server and Client clocks have some offset between them :) Let me show an example
- client sends {0, x, x} at C:0, S:5 (true offset is 5ms)
- server receives {0, 6, x} at C:1, S:6 (server is 1ms away)
- server sends {0, 6, 6} at C:1, S:6 (no processing time)
- client receives {0, 6, 6} at C:3, S:8 (reverse path is a bit longer)
So RTT = 3, ClockOffset = +/-4.5, CV = 3. Estimated clock offset is 4.5ms, but it's >= 3ms for sure.

Does it help? I'm unsure if I understand your question correctly :-)

Reply:
Thanks for putting up with all my questions.

After your most recent example, I understand better the meaning of "causality violation" and how it can be greater than zero in realistic scenarios.  Maybe the name "causality violation" is what is throwing me off, because it implies something that should be impossible.  There are no physical phenomena that violate the principle of causality. 

What this really is is a lower bound on the clock offset.  Once you know this value, how is it useful?  What can you do with it?